crates/sui-bcs-json/src/lib.rs

```diff
@@ -0,0 +1,841 @@
+            let ptr = vec.as_mut_ptr();
+            let len = vec.len();
+            // Prevent running `vec`'s destructor so we are in complete control
+            // of the allocation.
+            let _ = ManuallyDrop::new(vec);
+            unsafe {
+                *result = ptr as *mut u8;
+                *length = len as usize;
+            }
+            return 0;
+        }
+        Err(e) => {
+            ffi_helpers::update_last_error(e);
+            return 1;
+        }
+    }
+}
+
+/// Return 0 if the conversion from BCS to JSON is successful, and 1 or 2 for
+/// failure. 1 represents a failure from parsing the BCS to JSON, and 2
+/// represents an error building the CString from the JSON data.
+///
+/// The result argument will point to the address where the JSON
+/// string is stored. Make sure you release the allocated memory
+/// by calling sui_bcs_json_free_string function!
+#[no_mangle]
+pub extern "C" fn sui_bcs_to_json(
+    type_name: *const c_char,
+    bcs_buf_ptr: *mut *const u8,
+    len: usize,
+    result: *mut *const c_char,
+    pretty: bool,
+) -> usize {
+    let type_name = match unsafe { CStr::from_ptr(type_name) }.to_str() {
+        Ok(type_name) => type_name,
+        Err(e) => {
+            ffi_helpers::update_last_error(e);
+            return 1;
+        }
+    };
+
+    // we have a pointer to a pointer, thus we first get the address of the pointer
+    // to the bcs array data
+    // let ptr = unsafe { std::slice::from_raw_parts(bcs_buf_ptr, 1) };c
+    // and then reconstruct the slice
+    let mut data = unsafe { std::slice::from_raw_parts(*bcs_buf_ptr, len) };
+    match internal_bcs_to_json(type_name, &mut data, pretty) {
+        Ok(res) => {
+            let cstr = std::ffi::CString::new(res);
+            let cstr = match cstr {
+                Ok(c) => c.into_raw(),
+                Err(e) => {
+                    ffi_helpers::update_last_error(e);
+                    return 2;
+                }
+            };
+            unsafe { *result = cstr as *const i8 };
+            unsafe { *bcs_buf_ptr = data.as_ptr() }
+            return 0;
+        }
+        Err(e) => {
+            ffi_helpers::update_last_error(e);
+            return 1;
+        }
+    }
+}
+
+/// Frees a Rust-allocated Vec<u8>.
+#[no_mangle]
+pub extern "C" fn sui_bcs_json_free_array(ptr: *const u8, len: usize) {
+    unsafe {
+        // SAFETY: not going to mutate the value via the
+        drop(Vec::from_raw_parts(ptr as *mut u8, len, len));
+    }
+}
+
+/// Frees a Rust-allocated string.
+#[no_mangle]
+pub extern "C" fn sui_bcs_json_free_string(pointer: *const c_char) {
+    let _ = unsafe { std::ffi::CString::from_raw(pointer as *mut i8) };
+}
+
+/// Deserializes the data to given type T, and then serializes it as a JSON value
+fn to_json<T: Serialize + for<'a> Deserialize<'a>>(data: &mut &[u8]) -> Result<Value> {
+    let bcs: T = bcs::from_reader_no_check(data)?;
```

Comment:
This is a function that I added in my own `bcs` fork. The reason is that the original code https://github.com/zefchain/bcs/blob/main/src/de.rs#L89 calls `.end()?` which ensures that the BCS has the specified length and nothing else, which kinda defeats the purpose of being a reader.

Reply:
Please use the upstream bcs library and not a fork.

> which ensures that the BCS has the specified length and nothing else, which kinda defeats the purpose of being a reader.

This does not defeat the purpose and in fact is *required* for correctness. BCS is not a self-describing format and as such in order to use BCS properly you need to know beforehand the exact type of a serialized payload before you begin to deserialize it and you must fully and completely consume the entire serialized payload when deserializing to ensure that your local understanding of the type layout is correct otherwise you violate the determinism and canonical principles of the format. For example if there is a type T where two entities disagree about the layout we can run into problems:

```rust
struct Ta {
    field1: u64,
    field2: u8,
}

struct Tb {
    field1: u64,
    field2: u8,
    field3: u8,
}
```

If entity B serializes type T and sends it to entity A, if A doesn't exhaust the whole payload it'll end up with an incomplete view of the message that it was sent. This is a huge issue when it comes to cryptography and if this T was a transaction then entity A and B would disagree on the hash of the transaction. In general we need the serialization to fully round-trip everywhere.

```
payload == ser(des(payload))
```

Reply:
We were not planning on forking `bcs` but @stefan-mysten needed to do this to get things working, and we wanted to get your thoughts on what the best path is towards getting something that works for this case, and is palatable to include in upstream `bcs`.

Agreed that we cannot in general treat deserialization as "done" if there is data left in a payload, but the use case that this is trying to cater to is one where the C library does not know what the schema of the type is either, and is acting as a conduit for the host language (which does know) where the schema is known (e.g. for a Move type from some arbitrary package).

Reply:
As per our last discussion, reverted this change and fork deleted :D